scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/MultiSourceKeyGroupReader.java

```diff
@@ -0,0 +1,234 @@
+package org.apache.beam.sdk.extensions.smb;
+
+import java.io.ByteArrayInputStream;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.metrics.Distribution;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.transforms.join.CoGbkResultSchema;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Function;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.primitives.UnsignedBytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("unchecked")
+public class MultiSourceKeyGroupReader<FinalKeyT> {
+  private static final Logger LOG = LoggerFactory.getLogger(MultiSourceKeyGroupReader.class);
+
+  private enum AcceptKeyGroup {
+    ACCEPT,
+    REJECT,
+    UNSET
+  }
+
+  private Optional<KV<FinalKeyT, CoGbkResult>> head = null;
```

Comment:
overall, for Java code we've been preferring the plain null-check style rather than `Optional`s... it's just a bit less overhead and more streamlined.